### PR TITLE
Add first-class support for vertical orientation

### PR DESCRIPTION
**Pull request by valrus**
### Short description of changes:

This PR implements support for rendering the waveform in a vertical rather than horizontal orientation. To do this with as little disruption to the existing code as possible, I use two main approaches:
1. A utility object, `Orientation`, has two subclasses. `HorizontalOrientation` is basically a no-op. `VerticalOrientation` contains a mapping of "horizontally-oriented" attribute names to "vertically-oriented" ones (e.g. "width" becomes "height"). We can replace references to attributes in the DOM with calls to the `attrFor` method to ensure we use the correct attributes regardless of orientation, while preserving the "horizontally oriented" variable names in the code and without festooning the codebase with `if (vertical)` or some such.
2.  Actual canvas drawing is done by transformations on the context, so that we don't need to swap x's and y's every time we draw something in vertical mode.

### Breaking in the external API:

Nothing that I'm aware of. Default params all still assume horizontal orientation so none of their names have changed. This code caused no test breakages.

### Breaking changes in the internal API:

`CanvasEntry` now takes an `Orientation` parameter. This might be avoidable by having entries refer to the parent drawer's orientation instead, if need be.

### Todos/Notes:

I'm submitting this PR in an incomplete state because while I tried to keep it minimally disruptive, it does introduce an additional layer of indirection which might be a deal breaker to merge it. I wanted to see if the overall approach seemed mergeable before doing the following:

1. Porting more plugins. I ported the regions one since it's what I intend to use myself but I'd like to port the rest as well. It's a pretty mechanical process — just add an `attrFor` everywhere we touch the DOM — but a bit tedious.
2. Adding tests. Vertical orientation seems likely to cause regressions if future code changes omit an `attrFor` call somewhere, so tests are important. But I don't know the test framework so wanted to put that off until I know whether this might be merged.

### Related Issues and other PRs:

#634 is a request for this feature and the thread contains various hacks people used to get it sort-of working, mostly based on CSS transforms. AFAIK, all these approaches worked only partially at best, and they also have the problem of CSS transforms causing the waveform to be rendered perpendicular to the actual container.

### CODE REVIEW

File: example/vertical/index.html [hunk 1]
```diff
@@ -0,0 +1,124 @@
+<!DOCTYPE html>
+<html>
+    <head>
+        <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
+        <title>wavesurfer.js | Vertical </title>
+
```
Comment (thijstriemstra) [inline example/vertical/index.html:6]: I would use 'Vertical wave' or 'waveform' here

File: src/drawer.canvasentry.js [hunk 2]
```diff
@@ -13,7 +13,7 @@ import getId from './util/get-id';
  * render a waveform, depending on the zoom level.
  */
 export default class CanvasEntry {
-    constructor() {
+    constructor(orientation) {
```
Comment (thijstriemstra) [inline src/drawer.canvasentry.js:16]: your ticket description mentions no API breakage but  this new param breaks it.

File: src/drawer.canvasentry.js [hunk 3]
```diff
@@ -103,19 +105,21 @@ export default class CanvasEntry {
     updateDimensions(elementWidth, totalWidth, width, height) {
         // where the canvas starts and ends in the waveform, represented as a
         // decimal between 0 and 1
-        this.start = this.wave.offsetLeft / totalWidth || 0;
+        this.start = this.wave[this.orientation.attrFor('offsetLeft')] / totalWidth || 0;
```
Comment (thijstriemstra) [inline src/drawer.canvasentry.js:108]: I don't like this `attrFor` approach and would definitely prefer to keep the old object notation, e.g. `wave.offsetLeft`.

File: src/drawer.canvasentry.js [hunk 4]
```diff
@@ -128,32 +132,35 @@ export default class CanvasEntry {
...
     /**
-     * Set the fill styles for wave and progress
+     * Set the drawing context, including fill styles and transform,
+     * for wave and progress
      *
      * @param {string} waveColor Fill color for the wave canvas
      * @param {?string} progressColor Fill color for the progress canvas
      */
-    setFillStyles(waveColor, progressColor) {
+    setDrawingContext(waveColor, progressColor) {
```
Comment (thijstriemstra) [inline src/drawer.canvasentry.js:157]: This breaks the API. I'd rather keep the old setFillStyles around and introduce a new setDrawingContext-type method.

File: src/drawer.js [hunk 5]
```diff
@@ -29,6 +29,11 @@ export default class Drawer extends util.Observer {
          */
         this.height = params.height * this.params.pixelRatio;
 
+        this.orientation = util.makeOrientation(
```
Comment (thijstriemstra) [inline src/drawer.js:32]: ideally this has some comment with short explanation

File: src/wavesurfer.js [hunk 6]
```diff
@@ -313,6 +313,7 @@ export default class WaveSurfer extends util.Observer {
      */
     static create(params) {
         const wavesurfer = new WaveSurfer(params);
+        console.log("Hello from create");
```
Comment (thijstriemstra) [inline src/wavesurfer.js:316]: this can be removed

**Review by thijstriemstra — CHANGES_REQUESTED** [review]
thanks, really cool idea and good to have!

File: src/drawer.canvasentry.js (hunk 3 above)
Comment (valrus) [reply src/drawer.canvasentry.js:108]: TBH, I don't especially like it either, but I'm not sure how to do what you propose. We need some kind of shim to translate attributes such as this one for vertically oriented waveforms; can you suggest a way to do it more transparently, per your request? It seems like it would require overriding `wave` with some kind of wrapper, but this would be necessary for every other object for which we access the DOM attributes as well. Is that what you had in mind?

Comment (thijstriemstra) [reply src/drawer.canvasentry.js:108]: yea let's brainstorm about that. paging @sundayz @marizuccara

Comment (valrus) [reply src/drawer.canvasentry.js:108]: Perhaps a [Proxy object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy) would do the trick? I'll take a stab at it the next chance I have to work on this.

Comment (valrus) [reply src/drawer.canvasentry.js]: I just [checked in](https://github.com/katspaugh/wavesurfer.js/pull/2195/commits/117bbd4bfcfb41074866c01f45c6f89d61a9529e) a rough implementation of this but [Proxy support](https://caniuse.com/?search=proxy) may be prohibitively low. Babel flags it as undefined. I'm not sure what other options are but I may try a manual wrapper of some kind next.

Comment (thijstriemstra) [reply src/drawer.canvasentry.js]: huh, Proxy is supported in 99% of the browsers isn' it: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy#browser_compatibility

File: package.json [hunk 7]
```diff
@@ -29,6 +29,7 @@
     "lint-js-html": "lint-staged",
     "doc": "esdoc",
     "test": "karma start karma.conf.js",
+    "test:debug": "karma start karma.conf.js --browsers=Chrome --single-run=false",
```
Comment (thijstriemstra) [inline package.json:32]: maybe rename to  `test:chrome`?

File: src/drawer.canvasentry.js [hunk 8]
```diff
@@ -109,7 +109,9 @@ export default class CanvasEntry {
         // set wave canvas dimensions
         this.wave.width = width;
         this.wave.height = height;
-        let elementSize = { width: elementWidth + 'px' };
+        let elementSize = {
```
Comment (thijstriemstra) [inline src/drawer.canvasentry.js:112]: no need for this formatting

File: src/drawer.js [hunk 9]
```diff
@@ -86,39 +87,46 @@ export default class Drawer extends util.Observer {
...
+    getProgressPixels(wrapperBbox, clientX) {
+        if (this.params.rtl) {
+            return wrapperBbox.right - clientX;
+        } else {
+            return clientX - wrapperBbox.left;
+        }
+    }
+
     setupWrapperEvents() {
         this.wrapper.addEventListener('click', e => {
```
Comment (thijstriemstra) [inline src/drawer.js:121]: `e` might as well be called `event`

File: src/drawer.multicanvas.js [hunk 10]
```diff
@@ -82,6 +82,13 @@ export default class MultiCanvas extends Drawer {
          * @type {number}
          */
         this.barRadius = params.barRadius || 0;
+
+        /**
+         * Whether to render the waveform vertically.
+         *
+         * @type {boolean}
+         */
+        this.vertical = params.vertical;
```
Comment (thijstriemstra) [inline src/drawer.multicanvas.js:91]: docs should point out default is false

File: src/drawer.canvasentry.js (hunk 3 above)
Comment (valrus) [reply src/drawer.canvasentry.js]: Support seems decent in modern browsers but you would lose it for some older ones. I'm not sure if that's a concern. I would like to hash out a solution to this piece of things before addressing your other comments (which I'm happy to do) in case this turns out to be a hard blocker. Here's the options I see:
1. Implement this using Proxy objects, drop support entirely for older browsers as necessary. Not sure what to do about Babel flagging it in this case though.
2. Implement using Proxy objects but with a warning that vertical rendering is not supported in older browsers. Do a check in `orientation.js` whether `Proxy` is defined and, if not, make its functionality a no-op. Not sure if this will pass the linter.
3. Add a [Babel plugin](https://www.npmjs.com/package/babel-plugin-proxy) for a polyfill. It's 5 years old and warns "It is not suitable for production environments because performance impact is huge," so it may not be a _great_ option, but it is an option.
4. Something I'm missing??? I don't know if there's a way to do this cleanly without Proxy.

File: src/drawer.js (hunk 9 above)
Comment (valrus) [reply src/drawer.js:121]: `e` seems to be used everywhere else in the project. Happy to change this one but it will be inconsistent with what seems to be the convention.

File: src/util/capitalize.js [hunk 11]
```diff
@@ -0,0 +1,10 @@
+/**
+ * Capitalize a string's first character
+ *
+ * @param   {String} str String to capitalize
+ * @returns {String} str but with the first letter uppercased
+ * @example console.log(capitalize("hello")); // logs "Hello"
+ */
```
[the remark this answers is not in the document]
Comment (valrus) [reply src/util/capitalize.js:7]: Assuming I fix the API changes (besides the addition of the `vertical` param, obviously) should I assume this will be `@since 4.6.0`?

[the remark this answers is not in the document]
Comment (thijstriemstra) [reply src/util/capitalize.js:7]: correct

[the remark this answers is not in the document]
Comment (thijstriemstra) [reply src/util/capitalize.js:7]: outdated again, sorry, 4.7.0 would be better ;)

File: src/drawer.canvasentry.js (hunk 3 above)
Comment (thijstriemstra) [reply src/drawer.canvasentry.js]: > 1. drop support entirely for older browsers as necessary.

What browsers are we talking about exactly? and there's always a polyfill right?

> Do a check in orientation.js whether Proxy is defined and, if not, make its functionality a no-op. Not sure if this will pass the linter.

Sounds good.

Comment (thijstriemstra) [reply src/drawer.canvasentry.js]: ps. https://github.com/katspaugh/wavesurfer.js/blob/master/.eslintrc.js probably needs `Proxy` as well..?

File: src/drawer.canvasentry.js [hunk 12]
```diff
@@ -157,6 +157,21 @@ export default class CanvasEntry {
         }
     }
 
+    /**
+     * Set the canvas transforms for wave and progress
+     *
+     * @param {boolean} vertical Whether to render vertically
+     */
+    applyCanvasTransforms(vertical) {
+        if (vertical) {
```
Comment (thijstriemstra) [inline src/drawer.canvasentry.js:166]: add a comment for params passed to setTransform, it's black magic now ;)

File: src/plugin/regions/index.js [hunk 13]
```diff
@@ -127,11 +127,15 @@ export default class RegionsPlugin {
         });
         this.wavesurfer.Region = Region;
 
+        const scrollWidthProportion = 0.05;
```
Comment (thijstriemstra) [inline src/plugin/regions/index.js:130]: ideally, add comment why it's 0.05 or near that value

File: src/drawer.canvasentry.js (hunk 3 above)
Comment (sundayz) [reply src/drawer.canvasentry.js]: Pretty sure IE11 is the only browser that doesn't have Proxy support and can't be polyfilled

File: src/drawer.canvasentry.js (hunk 12 above)
Comment (valrus) [reply src/drawer.canvasentry.js:166]: Oops! I had one in a previous iteration but must have lost it in the shuffle.

File: src/plugin/regions/index.js (hunk 13 above)
Comment (valrus) [reply src/plugin/regions/index.js:130]: I didn't actually add that 0.05 to the code, just modified a line that contained it already, so I don't 100% know why it was used. But I can take a stab at it.

**Review by thijstriemstra — CHANGES_REQUESTED** [review]
Can you also add unit tests?

File: src/wavesurfer.js [hunk 14]
```diff
@@ -401,7 +401,11 @@ export default class WaveSurfer extends util.Observer {
         }
```
Comment (thijstriemstra) [inline src/wavesurfer.js:402]: the vertical parameter needs to be documented in this file, and added to default params

**Review by thijstriemstra — CHANGES_REQUESTED** [review]
great! one thing left.

File: src/wavesurfer.js [hunk 15]
```diff
@@ -115,6 +115,8 @@ import MediaElementWebAudio from './mediaelement-webaudio';
  * @property {boolean} splitChannels=false Render with separate waveforms for
  * the channels of the audio
  * @property {SplitChannelOptions} splitChannelsOptions={} Options for splitChannel rendering
+ * @property {boolean} vertical=false Render the waveform vertically instead of horizontally.
+ * This may not be compatible with any plugins besides the regions plugin.
```
Comment (thijstriemstra) [inline src/wavesurfer.js:119]: i would move that statement to changelog instead and add 'currently'

**Review by thijstriemstra — CHANGES_REQUESTED** [review]

File: src/wavesurfer.js (hunk 15 above)
Comment (thijstriemstra) [reply src/wavesurfer.js]: and build is failing on ci

Comment (valrus) [reply src/wavesurfer.js]: I don't see it. Where should I be looking?
Screen Shot 2021-03-14 at 2 58 53 PM

File: src/drawer.multicanvas.js [hunk 16]
```diff
@@ -210,11 +221,11 @@ export default class MultiCanvas extends Drawer {
         let lastEntry = this.canvases[this.canvases.length - 1];
 
         // wave
-        lastEntry.wave.parentElement.removeChild(lastEntry.wave);
+        lastEntry.wave.parentElement.removeChild(lastEntry.wave.proxiedElement);
 
         // progress
         if (this.hasProgressCanvas) {
-            lastEntry.progress.parentElement.removeChild(lastEntry.progress);
+            lastEntry.progress.parentElement.removeChild(lastEntry.progress.proxiedElement);
```
Comment (thijstriemstra) [inline src/drawer.multicanvas.js:228]: I'm still a little confused why you use `proxiedElement` property only here, and not in other places in multicanvas. Ideally we don't deal with this attribute at all..

Comment (valrus) [reply src/drawer.multicanvas.js:228]: I explain in this comment: https://github.com/katspaugh/wavesurfer.js/pull/2195/files#diff-c016f970dc3c8172ada63105abcd2b2f0e837a3c24c7592df6046dacd1bd4e0dR53

To expand a bit: Unfortunately, it doesn't seem possible to add/remove proxied elements to the DOM directly. [This blog post](https://hacks.mozilla.org/2015/07/es6-in-depth-proxies-and-reflect/) explains it briefly:

> There are limits to the fidelity of the illusion. You’ll find that `proxy !== target`. And a proxy will sometimes flunk type checks that the target would pass. Even if a proxy’s target is a DOM Element, for example, the proxy isn’t _really_ an Element; so something like `document.body.appendChild(proxy)` will fail with a `TypeError`.

I encountered this myself when testing this. So I have to use `proxiedElement` where a proxied element would touch the DOM directly.

Comment (thijstriemstra) [reply src/drawer.multicanvas.js:228]: I see. Perhaps `domElement` or `proxiedDOMElement` would better reflect the nature of the attribute. `proxiedElement` is too vague for me to be honest. I would also prefer to split the regions plugins changes into a separate PR that gets merged once this PR is merged. It would make it easier to read the changes made to core. thoughts?

File: CHANGES.md [hunk 17]
```diff
@@ -11,6 +11,8 @@ x.x.x (unreleased)
   - Fix mouseup not firing if click & drag a region handle & release outside browser window (#2213)
   - Added new `showTooltip` param allowing disabling region `title` tooltip (#2213)
 - Nullify `onaudioprocess` on remove to not execute in background (#2218)
+- Add new `vertical` param enabling displaying waveforms vertically; currently not compatible
```
Comment (thijstriemstra) [inline CHANGES.md:14]: the second part of the sentence can be used in the separate PR that adds regions plugin support. ps. are you planning on adding support in other plugins?

File: src/drawer.multicanvas.js (hunk 16 above)
Comment (valrus) [reply src/drawer.multicanvas.js]: Renamed to `domElement` per your request. The code you asked about should work, as the proxy passes through `set` to the underlying DOM element: https://github.com/katspaugh/wavesurfer.js/pull/2195/files#diff-c016f970dc3c8172ada63105abcd2b2f0e837a3c24c7592df6046dacd1bd4e0dR91

File: CHANGES.md (hunk 17 above)
Comment (valrus) [reply CHANGES.md]: I removed the regions plugin part of the PR as you requested and will submit it in a different PR, and updated CHANGES.md too. I should be able to add support for other plugins - I think it's mostly pretty straightforward. It's not immediately obvious to me what will have to be done for the svg stuff in the markers plugin but that might just be because I don't know svgs that well. I think I should be able to do it.

**Review by thijstriemstra — APPROVED** [review]
thanks @valrus for sticking through the review, looks good to me.

ps. can you add @valrus to the contributors/team @katspaugh?

**Comment by thijstriemstra** [comment]
@valrus if you could open the regions plugin PR in next 48 hours, we can include it in 4.7.0, otherwise it'll be in a next release.

**Comment by katspaugh** [comment]
@valrus I've sent you an invite, welcome! 🎉

File: src/drawer.canvasentry.js (hunk 3 above)
Comment (thijstriemstra) [reply src/drawer.canvasentry.js]: > I just [checked in](https://github.com/katspaugh/wavesurfer.js/pull/2195/commits/117bbd4bfcfb41074866c01f45c6f89d61a9529e) a rough implementation of this but [Proxy support](https://caniuse.com/?search=proxy) may be prohibitively low. Babel flags it as undefined. I'm not sure what other options are but I may try a manual wrapper of some kind next.
> Add a [Babel plugin](https://www.npmjs.com/package/babel-plugin-proxy) for a polyfill. It's 5 years old and warns "It is not suitable for production environments because performance impact is huge," so it may not be a great option, but it is an option.

I'm trying to reproduce this `Babel flags it as undefined` because babel-proxy-plugin `is 5 years old and warns "It is not suitable for production environments because performance impact is huge` which is problematic of course, stupid I missed that.

Comment (thijstriemstra) [reply src/drawer.canvasentry.js]: > Pretty sure IE11 is the only browser that doesn't have Proxy support and can't be polyfilled

Good point. I removed it and compilation doesn't fail. Let's see if the tests succeed as well and I'll remove it. See #2568
